### PR TITLE
Custom headers

### PR DESCRIPTION
**Pull request by adreed-msft**
Tackles issue #242.

Adds the following command line parameters:
- `--content-disposition`
- `--content-language`
- `--cache-control`

Tested by hand. If you want a unit test, I'm happy to write one but given this relies upon `azblob` and `azfile` functionality and is 99.9% new variables and setting them, I don't see the point.

### CODE REVIEW

File: cmd/copy.go [hunk 1]
```diff
@@ -1128,6 +1140,9 @@ func init() {
 	cpCmd.PersistentFlags().StringVar(&raw.metadata, "metadata", "", "upload to Azure Storage with these key-value pairs as metadata.")
 	cpCmd.PersistentFlags().StringVar(&raw.contentType, "content-type", "", "specifies content type of the file. Implies no-guess-mime-type.")
 	cpCmd.PersistentFlags().StringVar(&raw.contentEncoding, "content-encoding", "", "upload to Azure Storage using this content encoding.")
+	cpCmd.PersistentFlags().StringVar(&raw.contentDisposition, "content-disposition", "", "upload to Azure Storage using this content disposition.")
```
Comment (zezha-msft) [inline cmd/copy.go]: These headers are returned when downloading. Please update these messages to be more clear.

Refer to: https://docs.microsoft.com/en-us/rest/api/storageservices/put-blob

File: ste/JobPartPlan.go [hunk 2]
```diff
@@ -17,10 +17,13 @@ import (
 const DataSchemaVersion common.Version = 6
 
 const (
-	ContentTypeMaxBytes     = 256  // If > 65536, then jobPartPlanBlobData's ContentTypeLength's type  field must change
-	ContentEncodingMaxBytes = 256  // If > 65536, then jobPartPlanBlobData's ContentEncodingLength's type  field must change
-	MetadataMaxBytes        = 1000 // If > 65536, then jobPartPlanBlobData's MetadataLength field's type must change
-	BlobTierMaxBytes        = 10
+	ContentTypeMaxBytes        = 256  // If > 65536, then jobPartPlanBlobData's ContentTypeLength's type  field must change
```
Comment (zezha-msft) [inline ste/JobPartPlan.go]: minor:: maybe just use one variable? BlobCustomHeaderMaxBytes or something similar

**Comment by zezha-msft** [comment]
We should add unit tests for the ste module. But that can be done later.

File: ste/JobPartPlan.go (hunk 2 above)
Comment (adreed-msft) [reply ste/JobPartPlan.go]: I was mostly following the pattern I had noticed, but I will go ahead and change this.

Comment (adreed-msft) [reply ste/JobPartPlan.go]: Understandably though, the patterns of the old codebase (as we already know) were not very fantastic

File: cmd/copy.go [hunk 3]
```diff
@@ -688,13 +697,16 @@ func (cca *cookedCopyCmdArgs) processCopyJobPartOrders() (err error) {
 			BlockSizeInBytes:         cca.blockSize,
 			ContentType:              cca.contentType,
 			ContentEncoding:          cca.contentEncoding,
+			ContentLanguage:          cca.contentLanguage,
+			ContentDisposition:       cca.contentDisposition,
+			CacheControl:             cca.cacheControl,
```
Comment (JohnRusk) [inline cmd/copy.go]: Unlike the two new fields above it, this one appears to be never used.  Presumably the fact that it's not used is a bug?

File: ste/JobPartPlan.go [hunk 4]
```diff
@@ -17,10 +17,9 @@ import (
 const DataSchemaVersion common.Version = 6
 
 const (
-	ContentTypeMaxBytes     = 256  // If > 65536, then jobPartPlanBlobData's ContentTypeLength's type  field must change
-	ContentEncodingMaxBytes = 256  // If > 65536, then jobPartPlanBlobData's ContentEncodingLength's type  field must change
-	MetadataMaxBytes        = 1000 // If > 65536, then jobPartPlanBlobData's MetadataLength field's type must change
-	BlobTierMaxBytes        = 10
+	CustomHeaderMaxBytes = 256
```
Comment (JohnRusk) [inline ste/JobPartPlan.go]: Nice

File: ste/JobPartPlan.go [hunk 5]
```diff
@@ -17,10 +17,9 @@ import (
 const DataSchemaVersion common.Version = 6
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply ste/JobPartPlan.go]: Incremented the version

File: cmd/copy.go [hunk 6]
```diff
@@ -1126,8 +1138,11 @@ func init() {
 	cpCmd.PersistentFlags().StringVar(&raw.blockBlobTier, "block-blob-tier", "None", "upload block blob to Azure Storage using this blob tier.")
 	cpCmd.PersistentFlags().StringVar(&raw.pageBlobTier, "page-blob-tier", "None", "upload page blob to Azure Storage using this blob tier.")
 	cpCmd.PersistentFlags().StringVar(&raw.metadata, "metadata", "", "upload to Azure Storage with these key-value pairs as metadata.")
-	cpCmd.PersistentFlags().StringVar(&raw.contentType, "content-type", "", "specifies content type of the file. Implies no-guess-mime-type.")
-	cpCmd.PersistentFlags().StringVar(&raw.contentEncoding, "content-encoding", "", "upload to Azure Storage using this content encoding.")
+	cpCmd.PersistentFlags().StringVar(&raw.contentType, "content-type", "", "specifies content type of the file. Implies no-guess-mime-type. Returned on download of blob.")
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/copy.go]: Was following @zezha-msft's suggestion-- Willing to revert if necessary.

File: cmd/copy.go (hunk 3 above)
Comment (adreed-msft) [reply cmd/copy.go]: CacheControl _does_ make it into the headers in my experience. Happy to add internal tests though.

File: cmd/copy.go [hunk 7]
```diff
@@ -1126,8 +1138,11 @@ func init() {
...
+	cpCmd.PersistentFlags().StringVar(&raw.contentEncoding, "content-encoding", "", "upload to Azure Storage using this content encoding. Returned on download of blob.")
```
[the remark this answers is not in the document]
Comment (adreed-msft) [reply cmd/copy.go]: Not really. Changed the wording.

File: ste/JobPartPlan.go (hunk 4 above)
Comment (adreed-msft) [reply ste/JobPartPlan.go]: Thanks lol

File: cmd/copy.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (JohnRusk) [reply cmd/copy.go]: Changes you've made look good to me, and still retain Ze's suggestion, if I understand correctly. So I'll mark this conversation as resolved.

**Pull request merged**